mlir/tools/mlir-clang/Lib/clang-mlir.cc

```diff
@@ -770,6 +770,133 @@ MLIRScanner::VisitUnaryExprOrTypeTraitExpr(UnaryExprOrTypeTraitExpr *Uop) {
   }
 }
 
+bool isInAffineScope(Operation *op) {
+  auto *curOp = op;
+  while (auto *parentOp = curOp->getParentOp()) {
+    if (isa<mlir::AffineForOp>(parentOp))
+      return true;
+    curOp = parentOp;
+  }
+  return false;
+}
+
+bool hasAffineArith(Operation *op, AffineExpr &expr,
+                    mlir::Value &affineForIndVar) {
+  // skip IndexCastOp
+  if (isa<mlir::IndexCastOp>(op))
+    return hasAffineArith(op->getOperand(0).getDefiningOp(), expr,
+                          affineForIndVar);
+
+  // induction variable are modelled as memref<1xType>
```

Comment:
I don't really get this part. Is it saying that if an operation is a load from a 1-elem memref, and that value is stored from the index cast on an actual IV of an affine for, then any use of the result from that load should be treated as an IV? If so, where is this 1-elem memref pattern introduced?

Reply:
When we emit an ```affine.for``` we also generate the following code:
```
affine.for %arg0 = 0 to 1024
  %26 = index_cast %arg0: index to i32 // this is because in MLIR is of index type while is int for Clang in this case
  %27 = alloc() : memref<1xi32> //  from now on we use the induction as int type
  store %26, %27[0] : memref<1xi32>
 
  // more IR

 %33 = load %27[0]
 %34 = index_cast i32 to index // convert back to index type to use in MLIR
```
This code checks exactly this.